File: services/agentgraph-api/src/domain/projector.py
"""
Python port of @agentgraph/core projector — maps orchestrator events to AgentRun graph.
Mirrors the TypeScript implementation in packages/agentgraph-core/src/projector.ts.
"""
from __future__ import annotations

import time
from typing import Any

from .models import AgentEdge, AgentNode, AgentRun, PlanStep


class ProjectorState:
    def __init__(self) -> None:
        self.last_node_id: str | None = None
        self.pending_tool_call_id: str | None = None
        self.plan_node_id: str | None = None
        self.response_text: str = ""
        self.node_counter: int = 0

    def next_id(self, kind: str) -> str:
        self.node_counter += 1
        return f"{kind}-{self.node_counter}"


def _edge_id(source: str, target: str) -> str:
    return f"edge-{source}-{target}"


def _exec_edge(source: str, target: str, kind: str = "execution", animated: bool = False) -> AgentEdge:
    return AgentEdge(id=_edge_id(source, target), source=source, target=target, kind=kind, animated=animated)  # type: ignore[arg-type]


def project_events(events: list[dict[str, Any]], run_id: str, query: str) -> AgentRun:
    run = AgentRun(id=run_id, query=query, start_ms=time.time() * 1000)
    state = ProjectorState()
    now = time.time() * 1000

    def add_node(node: AgentNode, *, skip_edge: bool = False) -> None:
        run.nodes.append(node)
        if not skip_edge and state.last_node_id and node.kind != "tool_result":
            run.edges.append(_exec_edge(state.last_node_id, node.id))
        state.last_node_id = node.id

    def add_tool_result(node: AgentNode) -> None:
        run.nodes.append(node)
        if state.pending_tool_call_id:
            run.edges.append(_exec_edge(state.pending_tool_call_id, node.id, kind="data"))
            state.pending_tool_call_id = None
        elif state.last_node_id:
            run.edges.append(_exec_edge(state.last_node_id, node.id))
        state.last_node_id = node.id

    for event in events:
        evt_type = event.get("type", "")
        meta = event.get("metadata", {}) or {}

        if evt_type == "guardrail":
            section_raw = meta.get("section", "")
            section: str = "input" if "INPUT" in section_raw else "output"
            nid = state.next_id("guardrail")
            add_node(AgentNode(
                id=nid, kind="guardrail",
                label=f"{'Input' if section == 'input' else 'Output'} Guardrail",
                status="success" if meta.get("passed", True) else "error",
                start_ms=now,
                passed=meta.get("passed", True),
                score=meta.get("score"),
                guardrail_section=section,  # type: ignore[arg-type]
                violations=meta.get("violations", []),
            ))

        elif evt_type == "routing":
            nid = state.next_id("routing")
            add_node(AgentNode(
                id=nid, kind="routing",
                label=f"Route → {meta.get('routed_to', '')}",
                status="success", start_ms=now,
                model=meta.get("model"),
                model_tier=meta.get("model_tier"),
                routing_reason=meta.get("reason"),
                enabled_tools=meta.get("enabled_tools"),
            ))

        elif evt_type == "memory_recall":
            nid = state.next_id("memory_recall")
            add_node(AgentNode(
                id=nid, kind="memory_recall", label="Memory Recall",
                status="success", start_ms=now,
                total_memories=meta.get("total_memories"),
            ))

        elif evt_type == "planning":
            nid = state.next_id("planning")
            add_node(AgentNode(
                id=nid, kind="planning", label="Building Execution Plan",
                status="running", start_ms=now,
                model=meta.get("model"),
            ))

        elif evt_type == "plan_generated":
            nid = state.next_id("plan")
            state.plan_node_id = nid
            planning_node = next((n for n in reversed(run.nodes) if n.kind == "planning"), None)
            if planning_node:
                planning_node.status = "success"
                planning_node.end_ms = now
            steps = [
                PlanStep(
                    step_id=s["step_id"],
                    description=s["description"],
                    tool_name=s["tool_name"],
                    status="planned",
                )
                for s in meta.get("steps", [])
            ]
            add_node(AgentNode(
                id=nid, kind="plan",
                label=f"Plan: {meta.get('total_steps', 0)} steps",
                status="running", start_ms=now,
                plan_id=meta.get("plan_id"),
                goal=meta.get("goal"),
                steps=steps,
            ))

        elif evt_type in ("step_started", "step_verified", "step_failed") and state.plan_node_id:
            plan_node = next((n for n in run.nodes if n.id == state.plan_node_id), None)
            if plan_node and plan_node.steps:
                step_id = meta.get("step_id")
                for step in plan_node.steps:
                    if step.step_id == step_id:
                        if evt_type == "step_started":
                            step.status = "executing"
                        elif evt_type == "step_verified":
                            step.status = "completed"
                            step.duration_ms = meta.get("duration_ms")
                        elif evt_type == "step_failed":
                            step.status = "failed"
                            step.error = meta.get("error")
                            plan_node.status = "warning"
                        break

        elif evt_type == "thinking":
            section = meta.get("section", "")
            if section in ("UNDERSTANDING", "SYNTHESIS"):
                continue
            nid = state.next_id("thinking")
            add_node(AgentNode(
                id=nid, kind="thinking",
                label=event.get("content", section)[:60],
                status="warning" if section == "MODEL_FALLBACK" else "success",
                start_ms=now,
                section=section,
                icon=meta.get("icon"),
                reasoning=meta.get("reasoning"),
                failed_model=meta.get("failed_model"),
                fallback_model=meta.get("fallback_model"),
            ))

        elif evt_type == "tool_call":
            nid = state.next_id("tool_call")
            state.pending_tool_call_id = nid
            add_node(AgentNode(
                id=nid, kind="tool_call",
                label=meta.get("tool", "tool"),
                status="running", start_ms=now,
                tool=meta.get("tool"),
                tool_args=meta.get("arguments"),
                cache_hit=meta.get("cache_hit", False),
            ))

        elif evt_type == "knowledge_retrieval":
            tool_call = next((n for n in reversed(run.nodes) if n.kind == "tool_call"), None)
            if tool_call:
                tool_call.citations = meta.get("citations", [])
                tool_call.result_count = meta.get("result_count")

        elif evt_type == "tool_result":
            result = meta.get("result", {})
            has_error = bool(result.get("error"))
            tool_call = next((n for n in reversed(run.nodes) if n.kind == "tool_call" and n.status == "running"), None)
            if tool_call:
                tool_call.status = "error" if has_error else "success"
                tool_call.end_ms = now
                tool_call.duration_ms = result.get("duration_ms")
                tool_call.cache_hit = result.get("cache_hit")
                tool_call.error = result.get("error")
                state.pending_tool_call_id = tool_call.id
            nid = state.next_id("tool_result")
            add_tool_result(AgentNode(
                id=nid, kind="tool_result",
                label=f"{result.get('name', 'tool')} result",
                status="error" if has_error else "success",
                start_ms=now, end_ms=now,
                tool=result.get("name"),
                tool_output=result.get("output"),
                duration_ms=result.get("duration_ms"),
                cache_hit=result.get("cache_hit"),
                error=result.get("error"),
            ))

        elif evt_type == "reflecting":
            usage = meta.get("token_usage", {})
            nid = state.next_id("reflecting")
            add_node(AgentNode(
                id=nid, kind="reflecting", label="Token Reflection",
                status="success", start_ms=now,
                tokens={
                    "prompt": usage.get("prompt_tokens", 0),
                    "completion": usage.get("completion_tokens", 0),
                    "total": usage.get("total_tokens", 0),
                    "llm_calls": usage.get("llm_calls"),
                    "kb_chunks_retrieved": usage.get("kb_chunks_retrieved"),
                    "estimated_tokens_saved": usage.get("estimated_tokens_saved"),
                },
            ))

        elif evt_type == "human_escalation":
            iv = meta.get("intervention", {})
            nid = state.next_id("escalation")
            add_node(AgentNode(
                id=nid, kind="escalation", label="Human Approval Required",
                status="warning", start_ms=now,
                escalation_context=iv.get("context"),
                available_actions=iv.get("available_actions"),
                error=iv.get("error"),
            ))

        elif evt_type == "final_response_chunk":
            state.response_text += event.get("content", "")

        elif evt_type == "final_response":
            text = event.get("content") or state.response_text
            nid = state.next_id("response")
            add_node(AgentNode(
                id=nid, kind="response", label="Response",
                status="success", start_ms=now, end_ms=now,
                response_text=text[:500],
            ))
            run.status = "success"
            run.end_ms = now

        elif evt_type == "response_metadata":
            run.model = meta.get("model")
            run.confidence = meta.get("confidence")
            if meta.get("scope"):
                run.hki_domain = meta.get("scope")
            # Back-fill response node
            resp_node = next((n for n in reversed(run.nodes) if n.kind == "response"), None)
            if resp_node and meta.get("confidence") is not None:
                resp_node.confidence = meta["confidence"]

    return run

File: packages/myelin-core/src/projector.ts
import type {
  AgentEdge,
  AgentNode,
  AgentRun,
  EdgeKind,
  GraphMutation,
  NodeKind,
  NodeStatus,
  OrchestratorEvent,
  PlanStep,
} from "./types.js";

// ─── Projector state ─────────────────────────────────────────────────────────

export interface ProjectorState {
  lastNodeId: string | null;
  pendingToolCallId: string | null; // maps tool_call node id awaiting its tool_result
  planNodeId: string | null; // current plan node id for step updates
  responseText: string;
  nodeCounter: number;
}

export function createProjectorState(): ProjectorState {
  return {
    lastNodeId: null,
    pendingToolCallId: null,
    planNodeId: null,
    responseText: "",
    nodeCounter: 0,
  };
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

function nodeId(state: ProjectorState, kind: NodeKind): string {
  state.nodeCounter += 1;
  return `${kind}-${state.nodeCounter}`;
}

function edgeId(source: string, target: string): string {
  return `edge-${source}-${target}`;
}

function executionEdge(
  source: string,
  target: string,
  kind: EdgeKind = "execution",
  animated = false
): AgentEdge {
  return { id: edgeId(source, target), source, target, kind, animated };
}

function makeNode(
  id: string,
  kind: NodeKind,
  label: string,
  status: NodeStatus,
  startMs: number,
  extra: Partial<AgentNode> = {}
): AgentNode {
  return { id, kind, label, status, startMs, ...extra };
}

// ─── Core projector ───────────────────────────────────────────────────────────

export function projectEvent(
  run: AgentRun,
  event: OrchestratorEvent,
  state: ProjectorState,
  nowMs: number = Date.now()
): GraphMutation[] {
  const mutations: GraphMutation[] = [];

  function addNode(node: AgentNode): void {
    mutations.push({ op: "add_node", node });
    if (state.lastNodeId && node.kind !== "tool_result") {
      mutations.push({
        op: "add_edge",
        edge: executionEdge(state.lastNodeId, node.id),
      });
    }
    state.lastNodeId = node.id;
  }

  function addToolResultNode(node: AgentNode): void {
    mutations.push({ op: "add_node", node });
    if (state.pendingToolCallId) {
      // data edge: tool_call → tool_result
      mutations.push({
        op: "add_edge",
        edge: executionEdge(state.pendingToolCallId, node.id, "data"),
      });
      state.pendingToolCallId = null;
    } else if (state.lastNodeId) {
      mutations.push({
        op: "add_edge",
        edge: executionEdge(state.lastNodeId, node.id),
      });
    }
    state.lastNodeId = node.id;
  }

  switch (event.type) {
    case "guardrail": {
      const section =
        event.metadata.section === "INPUT_GUARDRAILS" ? "input" : "output";
      const label =
        section === "input" ? "Input Guardrail" : "Output Guardrail";
      const id = nodeId(state, "guardrail");
      addNode(
        makeNode(
          id,
          "guardrail",
          label,
          event.metadata.passed ? "success" : "error",
          nowMs,
          {
            passed: event.metadata.passed,
            score: event.metadata.score,
            violations: event.metadata.violations ?? [],
            guardrailSection: section,
          }
        )
      );
      break;
    }

    case "routing": {
      const id = nodeId(state, "routing");
      addNode(
        makeNode(
          id,
          "routing",
          `Route → ${event.metadata.routed_to}`,
          "success",
          nowMs,
          {
            model: event.metadata.model,
            modelTier: event.metadata.model_tier,
            routingReason: event.metadata.reason,
            enabledTools: event.metadata.enabled_tools,
          }
        )
      );
      break;
    }

    case "memory_recall": {
      const id = nodeId(state, "memory_recall");
      addNode(
        makeNode(id, "memory_recall", "Memory Recall", "success", nowMs, {
          totalMemories: event.metadata.total_memories,
        })
      );
      break;
    }

    case "planning": {
      const id = nodeId(state, "planning");
      addNode(
        makeNode(id, "planning", "Building Execution Plan", "running", nowMs, {
          section: event.metadata.section,
          model: event.metadata.model,
        })
      );
      break;
    }

    case "plan_generated": {
      const id = nodeId(state, "plan");
      state.planNodeId = id;
      const steps: PlanStep[] = event.metadata.steps.map(s => ({
        stepId: s.step_id,
        description: s.description,
        toolName: s.tool_name,
        status: "planned",
      }));
      // Update previous planning node to success if present
      const planningNode = run.nodes.findLast(n => n.kind === "planning");
      if (planningNode) {
        mutations.push({
          op: "update_node",
          id: planningNode.id,
          updates: { status: "success", endMs: nowMs },
        });
      }
      addNode(
        makeNode(
          id,
          "plan",
          `Plan: ${event.metadata.total_steps} steps`,
          "running",
          nowMs,
          {
            planId: event.metadata.plan_id,
            goal: event.metadata.goal,
            steps,
          }
        )
      );
      break;
    }

    case "step_started": {
      if (state.planNodeId) {
        const planNode = run.nodes.find(n => n.id === state.planNodeId);
        if (planNode?.steps) {
          const updated = planNode.steps.map(s =>
            s.stepId === event.metadata.step_id
              ? { ...s, status: "executing" as const }
              : s
          );
          mutations.push({
            op: "update_node",
            id: state.planNodeId,
            updates: { steps: updated },
          });
        }
      }
      break;
    }

    case "step_verified": {
      if (state.planNodeId) {
        const planNode = run.nodes.find(n => n.id === state.planNodeId);
        if (planNode?.steps) {
          const updated = planNode.steps.map(s =>
            s.stepId === event.metadata.step_id
              ? {
                  ...s,
                  status: "completed" as const,
                  durationMs: event.metadata.duration_ms,
                }
              : s
          );
          mutations.push({
            op: "update_node",
            id: state.planNodeId,
            updates: { steps: updated },
          });
        }
      }
      break;
    }

    case "step_failed": {
      if (state.planNodeId) {
        const planNode = run.nodes.find(n => n.id === state.planNodeId);
        if (planNode?.steps) {
          const updated = planNode.steps.map(s =>
            s.stepId === event.metadata.step_id
              ? { ...s, status: "failed" as const, error: event.metadata.error }
              : s
          );
          mutations.push({
            op: "update_node",
            id: state.planNodeId,
            updates: { steps: updated, status: "warning" },
          });
        }
      }
      break;
    }

    case "thinking": {
      const meta = event.metadata;
      // Model fallback is a distinct visual node
      if (meta.section === "MODEL_FALLBACK") {
        const id = nodeId(state, "thinking");
        addNode(
          makeNode(id, "thinking", "Model Fallback", "warning", nowMs, {
            section: meta.section,
            icon: meta.icon,
            reasoning: meta.reasoning,
            failedModel: meta.failed_model,
            fallbackModel: meta.fallback_model,
          })
        );
        break;
      }
      // Skip lightweight UNDERSTANDING/SYNTHESIS thinking nodes — they're noise unless planning is off
      if (meta.section === "UNDERSTANDING" || meta.section === "SYNTHESIS") {
        break;
      }
      const id = nodeId(state, "thinking");
      addNode(
        makeNode(
          id,
          "thinking",
          event.content?.slice(0, 60) ?? meta.section,
          "success",
          nowMs,
          {
            section: meta.section,
            icon: meta.icon,
            reasoning: meta.reasoning,
          }
        )
      );
      break;
    }

    case "tool_call": {
      const id = nodeId(state, "tool_call");
      state.pendingToolCallId = id;
      addNode(
        makeNode(id, "tool_call", event.metadata.tool, "running", nowMs, {
          tool: event.metadata.tool,
          toolArgs: event.metadata.arguments,
          cacheHit: event.metadata.cache_hit ?? false,
        })
      );
      break;
    }

    case "knowledge_retrieval": {
      // Enrich the pending tool_call node with KB metadata rather than add a new node
      const toolCallNode = run.nodes.findLast(n => n.kind === "tool_call");
      if (toolCallNode) {
        mutations.push({
          op: "update_node",
          id: toolCallNode.id,
          updates: {
            citations: event.metadata.citations,
            resultCount: event.metadata.result_count,
          },
        });
      }
      break;
    }

    case "tool_result": {
      const result = event.metadata.result;
      const hasError = Boolean(result.error);
      const id = nodeId(state, "tool_result");

      // Complete the tool_call node
      const toolCallNode = run.nodes.findLast(
        n => n.kind === "tool_call" && n.status === "running"
      );
      if (toolCallNode) {
        mutations.push({
          op: "update_node",
          id: toolCallNode.id,
          updates: {
            status: hasError ? "error" : "success",
            endMs: nowMs,
            durationMs: result.duration_ms,
            cacheHit: result.cache_hit,
            error: result.error,
          },
        });
        state.pendingToolCallId = toolCallNode.id;
      }

      addToolResultNode(
        makeNode(
          id,
          "tool_result",
          `${result.name} result`,
          hasError ? "error" : "success",
          nowMs,
          {
            tool: result.name,
            toolOutput: result.output,
            durationMs: result.duration_ms,
            cacheHit: result.cache_hit,
            error: result.error,
            endMs: nowMs,
          }
        )
      );
      break;
    }

    case "reflecting": {
      const usage = event.metadata.token_usage;
      const id = nodeId(state, "reflecting");
      addNode(
        makeNode(id, "reflecting", "Token Reflection", "success", nowMs, {
          tokens: {
            prompt: usage.prompt_tokens,
            completion: usage.completion_tokens,
            total: usage.total_tokens,
            llmCalls: usage.llm_calls,
            kbChunksRetrieved: usage.kb_chunks_retrieved,
            estimatedTokensSaved: usage.estimated_tokens_saved,
          },
        })
      );
      break;
    }

    case "human_escalation": {
      const iv = event.metadata.intervention;
      const id = nodeId(state, "escalation");
      addNode(
        makeNode(
          id,
          "escalation",
          "Human Approval Required",
          "warning",
          nowMs,
          {
            escalationContext: iv.context,
            availableActions: iv.available_actions,
            error: iv.error,
          }
        )
      );
      break;
    }

    case "final_response_chunk": {
      state.responseText += event.content ?? "";
      break;
    }

    case "final_response": {
      const text = event.content ?? state.responseText;
      const id = nodeId(state, "response");
      addNode(
        makeNode(id, "response", "Response", "success", nowMs, {
          responseText: text.slice(0, 500),
          endMs: nowMs,
        })
      );
      mutations.push({
        op: "update_run",
        updates: { endMs: nowMs, status: "success" },
      });
      break;
    }

    case "response_metadata": {
      mutations.push({
        op: "update_run",
        updates: {
          model: event.metadata.model,
          confidence: event.metadata.confidence,
          hkiDomain: event.metadata.scope as string | undefined,
        },
      });
      // Back-fill response node confidence
      const responseNode = run.nodes.findLast(n => n.kind === "response");
      if (responseNode && event.metadata.confidence != null) {
        mutations.push({
          op: "update_node",
          id: responseNode.id,
          updates: { confidence: event.metadata.confidence as number },
        });
      }
      break;
    }

    case "prompt_stack":
      // Internal — skip
      break;

    default:
      break;
  }

  return mutations;
}

// ─── Apply mutations to a run ─────────────────────────────────────────────────

export function applyMutations(
  run: AgentRun,
  mutations: GraphMutation[]
): AgentRun {
  let result = { ...run, nodes: [...run.nodes], edges: [...run.edges] };

  for (const mutation of mutations) {
    switch (mutation.op) {
      case "add_node":
        result.nodes = [...result.nodes, mutation.node];
        break;

      case "update_node": {
        result.nodes = result.nodes.map(n =>
          n.id === mutation.id ? { ...n, ...mutation.updates } : n
        );
        break;
      }

      case "add_edge":
        result.edges = [...result.edges, mutation.edge];
        break;

      case "update_run":
        result = { ...result, ...mutation.updates };
        break;
    }
  }

  return result;
}

// ─── Full replay ──────────────────────────────────────────────────────────────

export function replayEvents(
  runId: string,
  query: string,
  events: OrchestratorEvent[],
  startMs = Date.now()
): AgentRun {
  let run: AgentRun = {
    id: runId,
    query,
    startMs,
    status: "running",
    nodes: [],
    edges: [],
  };
  const state = createProjectorState();

  for (const event of events) {
    const mutations = projectEvent(run, event, state);
    run = applyMutations(run, mutations);
  }

  return run;
}
